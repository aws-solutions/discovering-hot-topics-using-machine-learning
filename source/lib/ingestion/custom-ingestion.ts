#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import * as fs from 'fs';
import { S3ToEventBridgeToLambda } from '../s3-event-notification/s3-eventbridge-lambda';
import path = require('path');

export class CustomIngestion extends cdk.NestedStack {
    public readonly s3Bucket: s3.Bucket;

    constructor(scope: Construct, id: string, props: cdk.NestedStackProps) {
        super(scope, id, props);

        const _streamArn = new cdk.CfnParameter(this, 'StreamARN', {
            type: 'String',
            description:
                'The name of the stream where the records from custom ingestion should be published for analysis',
            allowedPattern: '^arn:\\S+:kinesis:\\S+:\\d{12}:stream/\\S+$',
            constraintDescription: 'Please provide the Amazon Kinesis Data Streams name'
        });

        const _accessLoggingBucket = new cdk.CfnParameter(this, 'S3AccessLoggingBucket', {
            type: 'String',
            description: 'The ARN of the access logging bucket that should be used for S3 upload bucket',
            allowedPattern: '^arn:\\S+:s3:\\S+:\\S+:*',
            constraintDescription: 'Please provide a valid ARN for the S3 access logging bucket'
        });

        const _integrationEventBus = new cdk.CfnParameter(this, 'IntegrationBus', {
            type: 'String',
            allowedPattern: '^arn:\\S+:events:\\S+:\\d{12}:event-bus/\\S+$',
            description:
                'The Event Bus to publish metadata information directly for storage and visualization. This metadata information' +
                ' does not process through the workflow engine (e.g. metadata about call analytics processing)'
        });

        const _metadataNS = new cdk.CfnParameter(this, 'MetadataNS', {
            type: 'String',
            constraintDescription: 'Please provide the namespace to used for metadata events published to event bus',
            description:
                'The namespace is used as part of the event pattern to route events to targets on the event bus'
        });

        const _stream = kinesis.Stream.fromStreamArn(this, 'PublishCommentsStream', _streamArn.valueAsString);
        const _loggingBucket = s3.Bucket.fromBucketArn(this, 'AccessLoggingBucket', _accessLoggingBucket.valueAsString);
        const _integrationBus = events.EventBus.fromEventBusArn(
            this,
            'AppIntegrationBus',
            _integrationEventBus.valueAsString
        );

        /*
        The following environment variables for lambda functiona is an example for excel based ingestion.
            let _lambdaEnv = {
                STREAM_NAME: _stream.streamName,
                INTEGRATION_BUS_NAME: _integrationBus.eventBusName,
                ID: "1",
                CREATED_DATE: "2",
                TEXT: "3",
                LANG: "4",
                ACCOUNT_NAME: "xlsx_file",
                PLATFORM: "custom"
            }
        Note the values of ID, CREATED_DATE, TEXT, and LANG are column numbers from which the value will be extracted. A sample excel file is also available
        at this path source/lambda/custom-ingestion/test/fixtures/mock_data_file.xlsx.
        */

        /*
        The following environment variables for lambda function are set up for Amazon Transcribe call analytics based ingestion of data.
        It can be used for any other JSON based ingestion as well. The source/lambda/custom-ingestion/file_processor module defines the available processor 
        types and how they can be invoked through source/lambda/custom-ingestion/file_processor/processor_factory.py
         */
        let _lambdaEnv = {
            STREAM_NAME: _stream.streamName,
            INTEGRATION_BUS_NAME: _integrationBus.eventBusName,
            PROCESSOR_TYPE: 'TRANSCRIBE_CALL_ANALYTICS',
            ID: 'GENERATE',
            CREATED_DATE: 'NOW',
            TEXT: 'Content',
            LANG: 'LanguageCode',
            ACCOUNT_NAME: 'call_analytics',
            SENTIMENT: 'Sentiment',
            NAMESPACE: _metadataNS.valueAsString,
            PLATFORM: 'customingestion',
            LIST_SELECTOR: 'Transcript'
        };

        const _s3ToEventBridgeToLambda = new S3ToEventBridgeToLambda(this, 'CustomIngestion', {
            lambdaFunctionProps: {
                runtime: lambda.Runtime.PYTHON_3_11,
                code: lambda.Code.fromAsset('lambda/ingestion-custom'),
                handler: 'lambda_function.handler',
                environment: _lambdaEnv,
                timeout: cdk.Duration.minutes(15),
                memorySize: 256
            },
            bucketProps: {
                versioned: false,
                serverAccessLogsBucket: _loggingBucket,
                serverAccessLogsPrefix: 'customingestion/'
            },
            s3LoggingBucket: _loggingBucket
        });
        this.s3Bucket = _s3ToEventBridgeToLambda.s3Bucket;

        _stream.grantWrite(_s3ToEventBridgeToLambda.lambdaFunction);
        _integrationBus.grantPutEventsTo(_s3ToEventBridgeToLambda.lambdaFunction);

        new iam.Policy(this, 'TagS3Object', {
            document: new iam.PolicyDocument({
                statements: [
                    new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        actions: ['s3:PutObjectTagging'],
                        resources: [`${_s3ToEventBridgeToLambda.s3Bucket.bucketArn}/*`]
                    })
                ]
            })
        }).attachToRole(_s3ToEventBridgeToLambda.lambdaFunction.role!);

        const s3PolicyUpdateRole = new iam.Role(this, 'BucketPolicyCustomResource', {
            assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
            inlinePolicies: {
                LambdaFunctionServiceRolePolicy: new iam.PolicyDocument({
                    statements: [
                        new iam.PolicyStatement({
                            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
                            resources: [
                                `arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`
                            ]
                        }),
                        new iam.PolicyStatement({
                            actions: ['s3:PutBucketPolicy', 's3:GetBucketPolicy', 's3:PutBucketPublicAccessBlock'],
                            resources: [_loggingBucket.bucketArn]
                        })
                    ]
                })
            }
        });

        const s3PolicyUpdate = new lambda.Function(this, 'UpdateS3BucketPolicy', {
            handler: 'index.handler',
            runtime: lambda.Runtime.PYTHON_3_11,
            code: lambda.Code.fromInline(fs.readFileSync(path.join(__dirname, 'index.py'), 'utf8')),
            role: s3PolicyUpdateRole,
            tracing: lambda.Tracing.ACTIVE,
            timeout: cdk.Duration.minutes(15)
        });

        // prettier-ignore
        const customResourcePolicyUpdate = new cdk.CustomResource(this, 'UpdateS3Policy', {// NOSONAR - CDK construct initialization 
            resourceType: 'Custom::UpdateS3Policy',
            serviceToken: s3PolicyUpdate.functionArn,
            properties: {
                LOGGING_BUCKET_NAME: _loggingBucket.bucketName,
                SOURCE_BUCKET_NAME: this.s3Bucket.bucketName
            }
        });

        new cdk.CfnOutput(this, 'S3BucketToUploadData', {
            value: this.s3Bucket.bucketArn,
            description: 'Bucket location to upload source files for ingestion'
        });

        (s3PolicyUpdate.node.defaultChild as lambda.CfnFunction).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W89',
                    'reason': 'This is not a rule for the general case, just for specific use cases/industries'
                },
                {
                    'id': 'W92',
                    'reason': 'Impossible for us to define the correct concurrency for clients'
                }
            ]
        });

        (
            s3PolicyUpdateRole.node.tryFindChild('DefaultPolicy')!.node.tryFindChild('Resource') as iam.CfnPolicy
        ).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W12',
                    'reason':
                        'The cloudwatch policy narrows the permissions to the specific resources to the extent possible'
                }
            ]
        });
    }
}
