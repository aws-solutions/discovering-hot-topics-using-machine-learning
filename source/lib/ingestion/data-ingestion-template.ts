#!/usr/bin/env node
/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/


import * as ddb from 'aws-cdk-lib/aws-dynamodb';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { EventbridgeToLambda } from '@aws-solutions-constructs/aws-eventbridge-lambda';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import * as defaults from '@aws-solutions-constructs/core';

export interface LambdaAndDynamoDBProps {
    /**
     * Properties for the lambda function that triggers the event containing config information
     * for the resource target to process
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps;
    /**
     * User provided instance of the lambda function that triggers the event containing config
     * information for the resource target to process. If this property is set, @functionProps
     * is ignored
     */
    readonly existingLambda?: lambda.Function
    /**
     * Properties for the dynamoDB table that contains config information for the lambda functions
     * to use. The table is not a requirement for creating the construct. If @tableProps and
     * @existingTable are not set, the table is not created
     */
    readonly tableProps?: ddb.TableProps;
    /**
     * User provided instance of instance of the DyanamoDB table, if this is set, the properties
     * provided in @tableProps is ignored. If @tableProps and @existingTable are not set, the table
     * is not created
     */
    readonly existingTable?: ddb.Table;
    /**
     * The key path in SSM from where the credentials are to be read for the lambda to execute the API
     */
    readonly credentialKeyPath?: string;
}

/**
 * Interface defintion for the trigger constructs for the messaging service. The primary
 * source of trigger is a lambda function to begin with but other AWS resources may be
 * considered to be added in the future
 */
export interface TriggerSourceProps extends LambdaAndDynamoDBProps {
}

/**
 * Interface definition for the targets to which event bus will route events based on defined rules
 */
export interface ResourceTargetProps extends LambdaAndDynamoDBProps {
}

/**
 * Properties for the Ingestion construct
 */
export interface DataIngestionTemplateProps {
    /**
     * Properties to define tirgger source
     */
    readonly source: TriggerSourceProps;
    /**
     * Properties to define the resource targets
     */
    readonly target: ResourceTargetProps;
    /**
     * User provided instance of event bus on AWS EventBridge. If this is set, @ingestionEventBusProps
     * is ignored
     */
    readonly existingIngestionEventBus?: events.EventBus;
    /**
     * Properties to create a custom event bus
     */
    readonly ingestionEventBusProps?: events.EventBusProps;
    /**
     * Properties to define Rules to route events to targets
     */
    readonly ingestionEventRuleProps: events.RuleProps;
}

/**
 * A data type created to allow multiple construct resources created within the function to be
 * returned to the caller
 */
interface LambdaDDBConstruct {
    lambdaFunction: lambda.Function,
    ddbTable?: ddb.Table
}

/**
 * Qualify/ categorize constructs to be part of source trigger or resource targets
 */
enum Qualifier {
    SOURCE = 'Source',
    TARGET = 'Target'
}

/**
 * This construct will create a custom event bus using @IngestionProps supplied. It also automatically creates a
 * DLQ and associates it with the event bus, if none exists.
 *
 * This construct also adds the role policy permissions required for the source trigger to publish messages
 * on the event bus and for the targets to be able to read from the event bus
 */
export class DataIngestionTemplate extends Construct {
    private _bus: events.EventBus;
    private _rule: events.Rule;
    private _lambda: { [qualifier: string]: lambda.Function } = {};
    private _dynamoDB: { [qualifier: string]: ddb.Table } = {};

    constructor(scope: Construct, id: string, props: DataIngestionTemplateProps) {
        super(scope, id);

        // start of building target/ query engine lambda, optional DDB, and add optional SSM policy (TARGET)
        const _eventTarget = this.buildCustomBusAndAddRules(props);
        this._lambda[Qualifier.TARGET] = _eventTarget.lambdaFunction;

        if (props.target.credentialKeyPath) {
            this._lambda[Qualifier.TARGET].addToRolePolicy(this.buildSSMPolicy(props.target.credentialKeyPath));
        }

        if (props.target.tableProps || props.target.existingTable) {
            const _target = this.buildLambdaAndDynamoDB({
                existingLambda: this._lambda[Qualifier.TARGET],
                tableProps: props.target.tableProps,
                existingTable: props.target.existingTable
            }, Qualifier.TARGET);

            this._lambda[Qualifier.TARGET] = _target.lambdaFunction
            if (_target.ddbTable) {
                this._dynamoDB[Qualifier.TARGET] = _target.ddbTable;
            }
        }
        // end of building target/ query engine lambda, optional DDB, and add optional SSM policy

        // start of building source trigger lambda and optional DDB (SOURCE)
        const _source = this.buildLambdaAndDynamoDB(props.source, Qualifier.SOURCE);
        _source.lambdaFunction.addEnvironment("EVENT_BUS_NAME", this._bus.eventBusName)
        this._lambda[Qualifier.SOURCE] = _source.lambdaFunction;
        if (_source.ddbTable) {
            this._dynamoDB[Qualifier.SOURCE] = _source.ddbTable;
        }
        this._lambda[Qualifier.SOURCE].addToRolePolicy(this.buildPutEventPolicy());
        // end of building source trigger and optional DDB
    }

    /**
     * This method builds a custom bus and adds Rules to route events to target resources
     *
     * @param props
     */
    private buildCustomBusAndAddRules(props: DataIngestionTemplateProps): EventbridgeToLambda {
        if (props.existingIngestionEventBus) {
            this._bus = props.existingIngestionEventBus;
        }
        else if (props.ingestionEventBusProps) {
            this._bus = new events.EventBus(this, 'IngestionBus', props.ingestionEventBusProps);
        } else {
            throw new Error('Either ingestionEventBusProps or existingIngestionEventBus has to be set. Both cannot be undefined');
        }

        const _target = new EventbridgeToLambda(this, 'RuleTargetLambda', {
            lambdaFunctionProps: props.target.lambdaFunctionProps,
            existingLambdaObj: props.target.existingLambda,
            eventRuleProps: {
                ...props.ingestionEventRuleProps,
                eventBus: this._bus,
                enabled: true
            }
        });

        this._rule = _target.eventsRule;
        return _target;
    }

    /**
     * This method builds the lambda function and optionally the DynamoDB table that stores state/ config
     * information based on the parameters passed
     *
     * @param props - properties to build lambda and DynamoDB
     * @param id - @Qualifier
     */
    private buildLambdaAndDynamoDB(props: LambdaAndDynamoDBProps, id: Qualifier): LambdaDDBConstruct {
        const _lambda = defaults.buildLambdaFunction(this, {
            existingLambdaObj: props.existingLambda,
            lambdaFunctionProps: props.lambdaFunctionProps
        });

        let _ddbTable: ddb.Table | undefined;
        if (props.tableProps || props.existingTable) {
            const lambdaToDDB = new LambdaToDynamoDB(this, id, {
                existingLambdaObj: _lambda,
                dynamoTableProps: props.tableProps,
                existingTableObj: props.existingTable,
                tableEnvironmentVariableName: `${id.toUpperCase()}_DDB_TABLE`
            });

            _ddbTable = lambdaToDDB.dynamoTable;
        }

        return {
            lambdaFunction: _lambda,
            ddbTable: _ddbTable
        };
    }

    /**
     * Create SSM policy to perform GetParameter on the key path provided
     *
     * @param keyPath
     */
    private buildSSMPolicy(keyPath: string): iam.PolicyStatement {
        return new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            resources: [`arn:${cdk.Aws.PARTITION}:ssm:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:parameter${keyPath}`],
            actions: ['ssm:GetParameter']
        });
    }

    private buildPutEventPolicy(): iam.PolicyStatement {
        return new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            resources: [this._bus.eventBusArn],
            actions: ['events:PutEvents']
        })
    }

    public get rule(): events.Rule {
        return this._rule;
    }

    public get bus(): events.EventBus {
        return this._bus;
    }

    public get sourceLambda(): lambda.Function {
        return this._lambda[Qualifier.SOURCE];
    }

    public get configTable(): ddb.Table {
        return this._dynamoDB[Qualifier.SOURCE];
    }

    public get targetLambda(): lambda.Function {
        return this._lambda[Qualifier.TARGET];
    }

    public get stateTable(): ddb.Table {
        return this._dynamoDB[Qualifier.TARGET];
    }
}
