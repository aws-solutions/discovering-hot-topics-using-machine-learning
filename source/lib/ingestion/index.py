import json
import logging
import os
import uuid

import boto3
import cfnresponse

RESOURCE_PROPERTIES = "ResourceProperties"
SOURCE_BUCKET_NAME = "SOURCE_BUCKET_NAME"
LOGGING_BUCKET_NAME = "LOGGING_BUCKET_NAME"
PHYSICAL_RESOURCE_ID = "PhysicalResourceId"
POLICY_SID = "customingestionAccessLogsPolicy"
STATEMENT = "Statement"

DEFAULT_LEVEL = "DEBUG"

s3 = boto3.client("s3")
logger = logging.getLogger(__name__)
logger.setLevel(os.environ.get("LOG_LEVEL", DEFAULT_LEVEL))


def handler(event, context):
    logger.debug(f"The event is {event}")
    physical_resource_id = None

    try:
        physical_resource_id = event.get(PHYSICAL_RESOURCE_ID, uuid.uuid4().hex[:8])

        if event["RequestType"] == "Create" or event["RequestType"] == "Update":
            source_bucket_name = event[RESOURCE_PROPERTIES][SOURCE_BUCKET_NAME]
            logging_bucket_name = event[RESOURCE_PROPERTIES][LOGGING_BUCKET_NAME]
            account_id = context.invoked_function_arn.split(":")[4]
            aws_partition = context.invoked_function_arn.split(":")[1]

            s3.put_public_access_block(
                Bucket=logging_bucket_name,
                PublicAccessBlockConfiguration={
                    "BlockPublicAcls": True,
                    "IgnorePublicAcls": True,
                    "BlockPublicPolicy": False,
                    "RestrictPublicBuckets": True,
                },
                ExpectedBucketOwner=account_id,
            )

            policy_string = s3.get_bucket_policy(Bucket=logging_bucket_name)["Policy"]
            bucket_policy = json.loads(policy_string)

            update_policy = True

            for statement in bucket_policy[STATEMENT]:
                logger.debug(f"Statement is {json.dumps(statement)}")
                if "Sid" in statement and statement["Sid"] == POLICY_SID:
                    if (
                        f"arn:{aws_partition}:s3:::{source_bucket_name}"
                        not in statement["Condition"]["ArnLike"]["aws:SourceArn"]
                    ):
                        logger.info("Bucket policy exists but with a different bucket, hence removing it")
                        bucket_policy[STATEMENT].remove(statement)
                        break
                    else:
                        logger.info("Bucket policy exists, hence update not required")
                        update_policy = False
                        break

            if update_policy:
                bucket_policy[STATEMENT].append(
                    {
                        "Action": "s3:PutObject",
                        "Condition": {
                            "ArnLike": {"aws:SourceArn": [f"arn:{aws_partition}:s3:::{source_bucket_name}"]},
                            "StringEquals": {"aws:SourceAccount": account_id},
                        },
                        "Effect": "Allow",
                        "Principal": {"Service": "logging.s3.amazonaws.com"},
                        "Resource": f"arn:{aws_partition}:s3:::{logging_bucket_name}/customingestion*",
                        "Sid": POLICY_SID,
                    }
                )

                s3.put_bucket_policy(Bucket=logging_bucket_name, Policy=json.dumps(bucket_policy))

            s3.put_public_access_block(
                Bucket=logging_bucket_name,
                PublicAccessBlockConfiguration={
                    "BlockPublicAcls": True,
                    "IgnorePublicAcls": True,
                    "BlockPublicPolicy": True,
                    "RestrictPublicBuckets": True,
                },
                ExpectedBucketOwner=account_id,
            )
            logger.info("Policy Updated successfully. Sending success response to CloudFormation")

        cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_resource_id)
    except Exception as ex:
        logger.error(f"Failed to update policy, error is {str(ex)}")
        cfnresponse.send(
            event, context, cfnresponse.FAILED, {}, physicalResourceId=physical_resource_id, reason=str(ex)
        )
