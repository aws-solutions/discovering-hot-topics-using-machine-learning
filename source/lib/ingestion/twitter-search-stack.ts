#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from 'aws-cdk-lib';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { DiscoveringHotTopicsStack } from '../discovering-hot-topics-stack';
import { FeedProducer } from './feed-producer-construct';

export interface TwitterSearchIngestionProps {
    readonly ingestFrequency: string;
    readonly queryParameter: string;
    readonly supportedLang: string; //comma separated values of language codes
    readonly credentialKeyPath: string;
}

export class TwitterSearchIngestion extends cdk.NestedStack {
    constructor(scope: Construct, id: string, props?: cdk.NestedStackProps) {
        super(scope, id, props);

        const solutionName = this.node.tryGetContext('solution_name');

        const _supportedLang = new cdk.CfnParameter(this, 'SupportedLang', {
            type: 'String',
            description:
                'The list of languages to query the Search API with. The super set of languages supported is driven by Amazon Translate.' +
                'For an latest list of languages, please refer to the Amazon Translate documentation https://docs.aws.amazon.com/translate/latest/dg/what-is.html#language-pairs',
            maxLength: 43,
            minLength: 2,
            allowedPattern: '^$|([a-z]{2}-[a-z]{2}|[a-z]{2})(,([a-z]{2}-[a-z]{2}|[a-z]{2}))*',
            constraintDescription:
                'Provide a list of comma separated language iso-code values, Example: de,en,es,it,pt,fr,ja,ko,zh-cn (no spaces after the comma). The input did not match the validation pattern.'
        });

        const _queryParam = new cdk.CfnParameter(this, 'QueryParameter', {
            type: 'String',
            description:
                'The query you would like to execute on twitter. For details of how write a query and use operators, please go to https://developer.twitter.com/en/docs/tweets/search/guides/standard-operators',
            minLength: 3,
            maxLength: 500,
            default: 'entertainment'
        });

        const _credentialKeyPathParam = new cdk.CfnParameter(this, 'SSMPathForCredentials', {
            type: 'String',
            default:
                '/discovering-hot-topics-using-machine-learning/discovering-hot-topics-using-machine-learning/twitter',
            allowedPattern: '^(?!\\s*$).+',
            description: 'The SSM parameter store path of key where the credentials are stored as encrypted string',
            constraintDescription: 'The SSM parameter store path cannot be empty'
        });

        const _ingestFreqParam = new cdk.CfnParameter(this, 'IngestQueryFrequency', {
            type: 'String',
            default: 'cron(0/5 * * * ? *)',
            description:
                'The frequency at which API calls will be made to twitter in a cron expression format. For detailed documentation on schedule expression rules, please refer https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html',
            allowedPattern: DiscoveringHotTopicsStack.cronRegex,
            constraintDescription:
                "Please provide a valid cron expression of the format 'cron(0/5 * * * ? *)'. For details on CloudWatch cron expressions, please refer the following link https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html"
        });

        const _streamArn = new cdk.CfnParameter(this, 'StreamARN', {
            type: 'String',
            description: 'The name of the stream where search comments should be published for analysis',
            allowedPattern: '^arn:\\S+:kinesis:\\S+:\\d{12}:stream/\\S+$',
            constraintDescription: 'Please provide the Amazon Kinesis Data Streams name'
        });

        const _stream = kinesis.Stream.fromStreamArn(this, 'PublishCommentsStream', _streamArn.valueAsString);
        const _credentialPath = _credentialKeyPathParam.valueAsString
            ? _credentialKeyPathParam.valueAsString
            : `${this.node.tryGetContext('solution_name')}/${cdk.Aws.STACK_NAME}/twitter`;

        new FeedProducer(this, 'TwitterSearchAPI', {
            functionProps: {
                timeout: cdk.Duration.minutes(10),
                runtime: lambda.Runtime.NODEJS_18_X,
                code: lambda.Code.fromAsset('lambda/ingestion-producer'),
                handler: 'index.handler',
                environment: {
                    STREAM_NAME: _stream.streamName,
                    SOLUTION_NAME: solutionName,
                    STACK_NAME: cdk.Aws.STACK_NAME,
                    SUPPORTED_LANG: _supportedLang.valueAsString,
                    QUERY_PARAM: _queryParam.valueAsString,
                    CAP_NUM_RECORD: '25',
                    QUERY_RESULT_TYPE: 'mixed', //options are mixed, recent or popular
                    TWEET_MODE: 'extended',
                    TWITTER_CREDENTIAL_KEY_PATH: _credentialPath
                },
                reservedConcurrentExecutions: 1, // adding to throttle executions of this lambda function
                memorySize: 512
            },
            stream: _stream,
            ingestFrequency: _ingestFreqParam.valueAsString,
            credentialKeyPath: _credentialPath
        });
    }
}
