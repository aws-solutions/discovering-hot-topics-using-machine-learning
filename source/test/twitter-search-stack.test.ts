/**********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Template } from 'aws-cdk-lib/assertions';
import * as events from 'aws-cdk-lib/aws-events';
import * as cdk from 'aws-cdk-lib';
import { TwitterSearchIngestion } from '../lib/ingestion/twitter-search-stack';

test ('Event Bus creation', () => {
    const stack = new cdk.Stack();

    const _twtitterSearchIngestion = new TwitterSearchIngestion(stack, 'TwitterSearch', {
        parameters: {
            'EventBus': new events.EventBus(stack, 'testBus').eventBusArn,
            'SuppertedLang': "en,es,fr",
            'QueryParameter': "fakequery",
            'SSMPathForCredentials': "/discovering-hot-topics-using-machine-learning/discovering-hot-topics-using-machine-learning/twitter",
            'IngestQueryFrequency': 'cron(0 5 * * ? *)'
        }
    });
});
