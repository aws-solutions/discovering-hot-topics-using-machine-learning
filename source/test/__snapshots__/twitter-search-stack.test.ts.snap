// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Bus creation 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deArtifactHash2AF6C8BC": Object {
      "Description": "Artifact hash for asset \\"92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2de\\"",
      "Type": "String",
    },
    "AssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deS3Bucket1A1B9794": Object {
      "Description": "S3 bucket for asset \\"92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2de\\"",
      "Type": "String",
    },
    "AssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deS3VersionKeyD1737891": Object {
      "Description": "S3 key for asset version \\"92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2de\\"",
      "Type": "String",
    },
    "AssetParameterse037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0ArtifactHash1711CE02": Object {
      "Description": "Artifact hash for asset \\"e037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0\\"",
      "Type": "String",
    },
    "AssetParameterse037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0S3Bucket1EA20610": Object {
      "Description": "S3 bucket for asset \\"e037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0\\"",
      "Type": "String",
    },
    "AssetParameterse037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0S3VersionKey27431261": Object {
      "Description": "S3 key for asset version \\"e037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TwitterSearchNestedStackTwitterSearchNestedStackResource6905E30B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "EventBus": Object {
            "Fn::GetAtt": Array [
              "testBus07EAAD3A",
              "Arn",
            ],
          },
          "IngestQueryFrequency": "cron(0 5 * * ? *)",
          "QueryParameter": "fakequery",
          "SSMPathForCredentials": "/discovering-hot-topics-using-machine-learning/discovering-hot-topics-using-machine-learning/twitter",
          "SuppertedLang": "en,es,fr",
          "referencetoAssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deS3Bucket1A1B9794Ref": Object {
            "Ref": "AssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deS3Bucket1A1B9794",
          },
          "referencetoAssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deS3VersionKeyD1737891Ref": Object {
            "Ref": "AssetParameters92bb767e2980540b32ef0d403906d1b57004ab840bbeba0bf32b063c26f7e2deS3VersionKeyD1737891",
          },
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParameterse037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0S3Bucket1EA20610",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameterse037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0S3VersionKey27431261",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameterse037fea94b42ea69f1522c11ecde509519b9c153a94f3b96f9c8d28a67bcd6c0S3VersionKey27431261",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "testBus07EAAD3A": Object {
      "Properties": Object {
        "Name": "testBus",
      },
      "Type": "AWS::Events::EventBus",
    },
  },
}
`;
